Speed up first reservable time calculation

matti-lamppu opened the pull request:
## 🛠️ Changelog
[//]: # "Describe the changes in this pull request here."

Speeds up first reservable time calculation in a two different ways:

1. Pre-calculates affecting reservations with a materialized view. The view should be updated with a scheduled task.

2. Run calculation for fewer units based on pagination arguments. Use a custom `QuerySet.iterator()` to fetch chunks of reservation units and their affecting reservations

The above optimizations achieve <1s fetching with our test data when calculating FRT, regardless of `showOnlyReservable` or any pagination args.

## 🧪 Test plan
[//]: # "Help your fellow reviewer and write a short description of the fastest way to test your changes."

- Automated tests

## 🚧 Dependencies
[//]: # "Is this PR dependent on other changes outside this repository? Describe the changes, or add links to them."
[//]: # "e.g. This PR breaks X and is waiting for frontend changes before merging."

- Setup a periodic job in Django to refresh AffectingTimeSpans in platta envs.

## 🎫 Tickets
[//]: # "This pull request resolves all or part of the following ticket(s)."

- [TILA-3436][TILA]


[TILA]: https://helsinkisolutionoffice.atlassian.net/browse/TILA-3436


[TILA-3436]: https://helsinkisolutionoffice.atlassian.net/browse/TILA-3436?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

matti-lamppu commented:
Idea: Maybe there should be a flag set in cache which indicates that the materialized view refresh is valid, in case the job fails for some reason. This would then force a refresh in the endpoint if the value is missing.

vergama commented on reservations/migrations/0073_affectingtimespan.py:
```diff
@@ -0,0 +1,121 @@
+                    "affected_reservation_unit_ids",
+                    django.contrib.postgres.fields.ArrayField(base_field=models.IntegerField(), size=None),
+                ),
+                ("is_reservable", models.BooleanField()),
```
Should this be `is_blocking`?

matti-lamppu replied:
Yes, Fixed this.

vergama commented on reservations/migrations/0073_affectingtimespan.py:
```diff
@@ -0,0 +1,121 @@
+                        res.buffer_time_before,
+                        res.buffer_time_after,
+                        res.is_blocking
+                    ORDER BY res.buffered_start_datetime, res.reservation_id;
```
Does it matter [that](https://www.postgresql.org/docs/current/sql-refreshmaterializedview.html):
> If there is an ORDER BY clause in the materialized view's defining query, the original contents of the materialized view will be ordered that way; but REFRESH MATERIALIZED VIEW does not guarantee to preserve that ordering.

Not because of `AffectingTimeSpan.Meta.ordering`?

matti-lamppu replied:
Oh, good to know. Still, I think this should be fine since AffectingTimeSpans are ordered chronologically when they are merged in `FirstReservableTimeHelper._get_elements_for_reservation_units`.

vergama commented:
> Idea: Maybe there should be a flag set in cache which indicates that the materialized view refresh is valid, in case the job fails for some reason. This would then force a refresh in the endpoint if the value is missing.

Or value of when the view was last updated; force refresh if too old?
Some complications may be caused by parallel endpoint hits encountering invalid mat.view, in which case the refresh should be synchronized...

vergama reviewed (approved):
"The dude abides."

matti-lamppu commented:
Added a value to cache for when the view was last updated. The FRT calculation will check this and refresh the view if it is too old. Refreshing the view with `AffectingTimeSpan.refresh()` updates the cached value.

Also, added error handling logic to the `.refresh()` methods on both materialized views to raise an error if the refresh fails during local development, but log to sentry in other environments.

matti-lamppu commented:
Merged #1234, #1274, #1261 to this PR so that they can be merged to main together.

matti-lamppu commented:
/ff
